### PR TITLE
Okapi 773 switch discovery only if all succeeds

### PR DESCRIPTION
**Pull request by adamdickmeiss**
This PR improves coverage in general for install. It ensures that the tenant module association is not happening until all modules have been completing tenant init and tenant permissions interface. Of course, if the modules do not implement these interfaces, they will not be considered. Also if a module is not running at all, of course, also that stops upgrading. There's no switch for this. This changes behavior for the better, so a desired breaking change = bug fix.

Out of scope of this issue:

We should extend the tenant interface with a rollback mechanism so that module can be notified to rollback if they have succeeded but another module has failed. 

Finally, we could do a pre-check tenant init which has no side effects but just let the module ack that it can do upgrade (without actually doing anything). https://issues.folio.org/browse/OKAPI-814

### CODE REVIEW

**Comment by skoczko** [comment]
@adamdickmeiss Can you document changed/additions to the Tenant "init" interface?

File: okapi-core/src/main/java/org/folio/okapi/managers/ProxyService.java [hunk 1]
```diff
@@ -1108,14 +1092,13 @@ private void authForSystemInterface(ModuleDescriptor authMod, RoutingEntry filt,
...
-        fut.handle(new Failure<>(res.getType(), res.cause()));
+        fut.handle(res);
         return;
       }
       OkapiClient cli = res.result();
       String deftok = cli.getRespHeaders().get(XOkapiHeaders.TOKEN);
-      logger.debug("authForSystemInterface:"
-        + Json.encode(cli.getRespHeaders().entries()));
+      logger.debug("authForSystemInterface: {}",
+        Json.encode(cli.getRespHeaders().entries()));
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/managers/ProxyService.java]:
```suggestion
        () -> Json.encode(cli.getRespHeaders().entries()));
```
Avoid encoding when debug logging is diabled.

File: okapi-core/src/main/java/org/folio/okapi/managers/ProxyService.java [hunk 2]
```diff
@@ -1033,7 +1017,7 @@ private DeploymentDescriptor pickInstance(List<DeploymentDescriptor> instances)
    */
   public void callSystemInterface(Tenant tenant, ModuleInstance inst,
     String request, ProxyContext pc,
-    Handler<ExtendedAsyncResult<OkapiClient>> fut) {
+    Handler<AsyncResult<OkapiClient>> fut) {
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/managers/ProxyService.java]: Good!

File: okapi-core/src/main/java/org/folio/okapi/MainVerticle.java [hunk 3]
```diff
@@ -457,7 +434,7 @@ private void checkSuperTenant(String okapiModule, Promise<Void> promise) {
       if (res.succeeded()) {
         logger.info("Deploy completed succesfully");
       } else {
-        logger.info("Deploy failed: {}", res.cause());
+        logger.info("Deploy failed", res.cause());
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/MainVerticle.java]: If it is intentional that the failure is only logged and not reported via promise there should be a `//` comment saying so.

File: okapi-core/src/main/java/org/folio/okapi/managers/PullManager.java [hunk 4]
```diff
@@ -52,17 +52,17 @@ private void getRemoteUrl(Iterator<String> it,
     final Buffer body = Buffer.buffer();
     HttpClientRequest req = httpClient.getAbs(url, res1 -> {
       if (res1.failed()) {
-        logger.info("pull for " + baseUrl + " failed with status "
-          + res1.cause().getMessage());
+        logger.warn("pull for {} failed with status {}", baseUrl,
+          res1.cause().getMessage());
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/managers/PullManager.java]:
```suggestion
        logger.warn("pull for {} failed: {}", baseUrl,
          res1.cause().getMessage(), res1);
```
Add stacktrace.

File: okapi-core/src/main/java/org/folio/okapi/MainVerticle.java (hunk 3 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/MainVerticle.java]: It should probably report failure, but I'd rather not change behavior WRT this, so I'll add a comment.

File: okapi-core/src/main/java/org/folio/okapi/managers/PullManager.java (hunk 4 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/managers/PullManager.java]: There's no stacktrace (not an exception)

File: okapi-core/src/test/java/org/folio/okapi/ModuleTest.java [hunk 5]
```diff
@@ -2681,32 +2683,97 @@ private void undeployFirst(Handler<AsyncResult<Void>> fut) {
...
-      conf.remove("mongo_db_init");
-      conf.remove("postgres_db_init");
+      DeploymentOptions opt = new DeploymentOptions().setConfig(conf);
+      vertx.deployVerticle(MainVerticle.class.getName(), opt, res -> {
+        context.assertTrue(res.succeeded());
+        async.complete();
+      });
+    });
+    async.await(1000);
+  }
```
Comment (julianladisch) [inline okapi-core/src/test/java/org/folio/okapi/ModuleTest.java]: The duplicate code should be extracted into a method:

```
 private void undeployFirstAndDeploy(TestContext context, JsonObject aConf, Handler<AsyncResult<String>> fut) {
   Async myAsync = context.async();
   undeployFirst(context.asyncAssertSuccess(handler -> {
     DeploymentOptions opt = new DeploymentOptions().setConfig(aConf);
     vertx.deployVerticle(MainVerticle.class.getName(), opt, res -> {
       fut.handle(res);
       myAsync.complete();
     });
   }));
   myAsync.await(2000);
 }

 private void undeployFirstAndDeploy(TestContext context, JsonObject aConf) {
   undeployFirstAndDeploy(context, aConf, context.asyncAssertSuccess());
 }
```
Example usage:
```
undeployFirstAndDeploy(context, conf);
```
```
undeployFirstAndDeploy(context, conf, context.asyncAssertFailure(cause -> {
  Assert.assertThat(cause.getMessage(), containsString(
      "password authentication failed for user \"okapi\""));
}));
```

File: okapi-core/src/main/java/org/folio/okapi/MainVerticle.java (hunk 3 above)
Comment (julianladisch) [reply okapi-core/src/main/java/org/folio/okapi/MainVerticle.java]: Created an issue for this: https://issues.folio.org/browse/OKAPI-816

File: okapi-core/src/test/java/org/folio/okapi/ModuleTest.java (hunk 5 above)
Comment (adamdickmeiss) [reply okapi-core/src/test/java/org/folio/okapi/ModuleTest.java]: ok. Done

**Pull request merged**